### PR TITLE
Add caching modules and dependencies

### PR DESCRIPTION
**Pull request by jackbackes**
- Added Moka and Redis dependencies in Cargo.toml
- Created caching module with submodules for Moka, Redis, and NoOp caches
- Implemented caching traits and asynchronous operations for cache types

### CODE REVIEW

File: src/extractor_router.rs [hunk 1]
```diff
@@ -13,6 +13,8 @@ pub struct ExtractorRouter {
     coordinator_client: Arc<CoordinatorClient>,
 }
 
+
```
Comment (diptanu) [inline src/extractor_router.rs:16]: Are these from formatting?

File: src/caching/redis_cache.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        let key = serde_json::to_string(key)?;
+        redis::cmd("DEL")
+            .arg(key)
+            .query_async(&mut conn)
+            .await?;
+        Ok(())
+    }
+}
+
+// TODO: integrate Redis into local dev setup so we can run this test
```
Comment (diptanu) [inline src/caching/redis_cache.rs:69]: Should be super easy, can we just add the dependency in Makefile and uncomment these before merging?

**Review by diptanu — CHANGES_REQUESTED** [review]
Minor comments, looks good overall. I am not very happy with encoding the key as json but I understand why it had to be done. I wonder if there is a faster serialization/de-serilazation. I wonder if Avro would be a better fit than json for this use case, we can still use serde to move back and forth between the structs and serialized forms - https://docs.rs/avro-rs/latest/avro_rs/

File: src/extractor_router.rs (hunk 1 above)
Comment (jackbackes) [reply src/extractor_router.rs]: Ran make fmt so it has standard formatting now.

File: src/caching/redis_cache.rs (hunk 2 above)
Comment (jackbackes) [reply src/caching/redis_cache.rs:67]: Added Redis. Test now passes if Redis is running!

**Pull request merged**